src/ucp/wireup/select.c

```diff
@@ -1532,12 +1552,177 @@ static double ucp_wireup_get_lane_bw(ucp_worker_h worker,
     return ucs_min(bw_local, bw_remote);
 }
 
+static int
+ucp_proto_select_info_score_rcompare(const void *e1, const void *e2)
+{
+    const ucp_wireup_select_info_t *info1 = e1;
+    const ucp_wireup_select_info_t *info2 = e2;
+
+    /* reverse the value of ucp_score_cmp to sort scores in descending order */
+    return -ucp_score_prio_cmp(info1->score, info1->priority,
+                               info2->score, info2->priority);
+}
+
+static unsigned
+ucp_wireup_add_fast_lanes_a2a(ucp_worker_h worker,
+                              ucp_proto_select_info_array_t *sinfo_array,
+                              const ucp_wireup_select_params_t *select_params,
+                              ucp_lane_type_t lane_type, unsigned max_lanes,
+                              ucp_wireup_select_context_t *select_ctx)
+{
+    ucp_lane_index_t num_lanes = 0;
+    double max_bw              = 0;
+    double lane_bw             = 0;
+    ucp_context_h context      = worker->context;
+    const double max_ratio     = 1. / context->config.ext.multi_lane_max_ratio;
+    ucs_status_t status;
+    const ucp_wireup_select_info_t *sinfo;
+
+    qsort(sinfo_array->buffer, sinfo_array->length,
+          sizeof(ucp_wireup_select_info_t),
+          ucp_proto_select_info_score_rcompare);
+
```

Comment:
is it possible to merge this function with `ucp_wireup_add_fast_lanes_d2d`?
To prevent code duplication

Reply:
I don't think so. We need to keep pairwise (d2d) logic for compatibility reason, it adds lanes unordered only by BW filter, in case of a2a we need sort them and cut by max_lanes to guarantee the best and stable choice.